### ISSUE TITLE
'$' sign in the output base make all tests failed on windows

### ISSUE THREAD

**Issue by damienmg** [bazelbuild/bazel#2306]
See bazelbuild/continuous-integration#43 for more information

Categorizing as breakage since it prevents correct vetting of pull requests

**Comment by damienmg** [comment]
Likely culprit: https://github.com/bazelbuild/bazel/commit/af1e3c27364b5b97facd1148dd7d59a4c812f81b because of https://github.com/bazelbuild/bazel/blame/master/tools/test/test-setup.sh#L104

**Comment by damienmg** [comment]
Humm af1e3c2 is not in 0.4.2 that was already having the problem

**Comment by damienmg** [comment]
breakage started with 0.4.2

**Comment by damienmg** [comment]
Ok nothing to do with af1e3c2, the problems come from the generated .cmd file (tested by adding -x to the test-setup.sh)

**Comment by damienmg** [comment]
l29 of the cmd file, there is missing quotes, adding them fix the problem

**Comment by damienmg** [comment]
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template_windows.txt#L29

So at this line, the %%i should be quoted to fix the issue.

Since the issue needs another release and a proper regression test (set output_base to something with a dollar sign and execute a java test, useful integration test for Bazel on other platform too), reassigning to a release blocker.

Side note: we do not run any shell tests for windows.

**Comment by damienmg** [comment]
And Laszlo is the proud owner of that bug since he wrote the initial template :)

**Comment by damienmg** [comment]
(side note: since the commit with the bug was before, I suspect another commit changed the execroot computation and triggered the bug)

**Comment by laszlocsomor** [comment]
Blocked by https://github.com/bazelbuild/bazel/issues/2323

**Comment by laszlocsomor** [comment]
Culprit fixed, now looking at this error.

**Comment by laszlocsomor** [comment]
Are you sure this wasn't just a transient test flakiness?

Here, I built Bazel at ccb78ec8 then ran the tests with it at the same revision:
```
$ /c/tempdir/bazel_tmp.exe --batch --bazelrc=c:/users/laszlocsomor/.bazelrc --output_user_root=/c/tmp2/a\$bc test src/test/java/com/google/devtools/build/lib:windows-tests
```

This works for me. (The bazelrc part is a workaround for #2323).

**Comment by damienmg** [comment]
See the original bug on continuous integration, this is definitely an
permanent issue. Quoting after the set did the fix for me. Is it related to
running the test in msys vs from CMD.exe?

**Comment by laszlocsomor** [comment]
Ok, I'll take another look, but so far I haven't been able to repro this.

I'm running the test from msys, did you try from cmd.exe? How did you get around `PATH` not being defined? (Because `Path` is defined, and envvars are case-insensitive, see #1842, and I can't undef an envvar.)

**Comment by laszlocsomor** [comment]
Ah, never mind. I can `set path=` then `set PATH=.....`

**Comment by laszlocsomor** [comment]
I could repro, investigating.

**Comment by laszlocsomor** [comment]
@damienmg , you're right, with `--output_user_root=/c/tmp2/a\$bc` the `$bc` interpreted as a shell variable, but that's not the only problem: msys and/or the launcher script think that `\` is a directory separator, not an escaping character, so `a\$bc` is treated as a path with `a` being the parent, `$bc` the child.

**Comment by laszlocsomor** [comment]
This shouldn't happen since https://github.com/bazelbuild/bazel/commit/b64349e91debe8abb8a218c6e8888dc9b42fdea3 unless the user specifically provides such a directory.

**Comment by dslomov** [comment]
Fixed, right?

**Comment by laszlocsomor** [comment]
Not really; if the user specifies `--output_user_root=/c/bad\$dir` then we still have the problem I think, but at least bazel itself won't generate such directory names. So I think we're OK.

**Comment by damienmg** [comment]
Reopen as non release blocker and lower priority then (P3?)?

**Comment by lberki** [comment]
There is also potentially a problem with other special characters (backslash, space, octothorpe and others). Given that they need to be passed through a bash script and cmd.exe, they'll need close scrutiny.

**Comment by laszlocsomor** [comment]
- You can't create directories on Windows with backslash (\) in their name. Not even from under MSYS.
- You can create directories with double quote (") from under MSYS, and you can cd into them from cmd.exe, though the (") is rendered as an invalid character.
- You can create directories with space in their name, that is fully supported, see e.g. "Program Files"
- I think other invalid characters are also unsupported, but I can write a small test to try them.

**Comment by lberki** [comment]
It's fixed with #2342 in b64349e .

**Comment by laszlocsomor** [comment]
- hashmark (for the fun-loving: hashtag) is fully supported

**Comment by laszlocsomor** [comment]
This has been fixed for a while now, we should no longer create output roots with "$" in the name.
